OsmAnd-java/src/main/java/net/osmand/router/BinaryRoutePlanner.java

```diff
@@ -747,64 +725,80 @@ protected void processRestriction(RoutingContext ctx, RouteSegment inputNext, bo
 		}
 	}
 
-
 	private RouteSegment processIntersections(RoutingContext ctx, PriorityQueue<RouteSegment> graphSegments,
-			TLongObjectHashMap<RouteSegment> visitedSegments,  float distFromStart, RouteSegment segment,
-			short segmentPoint, RouteSegment inputNext, boolean reverseWaySearch, boolean doNotAddIntersections, 
-			boolean[] processFurther) {
-		boolean thereAreRestrictions;
-		RouteSegment itself = null;
-		processFurther[0] = true;
-		Iterator<RouteSegment> nextIterator = null;
-		if (inputNext != null && inputNext.getRoad().getId() == segment.getRoad().getId() && inputNext.next == null) {
-			thereAreRestrictions = false;
-		} else {
-			thereAreRestrictions = proccessRestrictions(ctx, segment, inputNext, reverseWaySearch);
-			if (thereAreRestrictions) {
-				nextIterator = ctx.segmentsToVisitPrescripted.iterator();
-				if (TRACE_ROUTING) {
-					println("  >> There are restrictions");
+			TLongObjectHashMap<RouteSegment> visitedSegments,  RouteSegment currentSegment,
+			boolean reverseWaySearch, boolean doNotAddIntersections) {
+		RouteSegment nextCurrentSegment = null;
+		int targetEndX = reverseWaySearch ? ctx.startX : ctx.targetX;
+		int targetEndY = reverseWaySearch ? ctx.startY : ctx.targetY;
+		final int x = currentSegment.getRoad().getPoint31XTile(currentSegment.getSegmentEnd());
+		final int y = currentSegment.getRoad().getPoint31YTile(currentSegment.getSegmentEnd());
+		float distanceToEnd = h(ctx, x, y, targetEndX, targetEndY);
+		// reassign @distanceToEnd to make it correct for visited segment
+		currentSegment.distanceToEnd = distanceToEnd; 
+		
+		final RouteSegment connectedNextSegment = ctx.loadRouteSegment(x, y, ctx.config.memoryLimitation - ctx.memoryOverhead, reverseWaySearch);
+		RouteSegment roadIter = connectedNextSegment;
+		boolean directionAllowed = true;
+		boolean singleRoad = true;
+		while (roadIter != null) {
+			if (currentSegment.getSegmentEnd() == roadIter.getSegmentStart() && roadIter.road.getId() == currentSegment.getRoad().getId() ) {
+				nextCurrentSegment = roadIter.initRouteSegment(currentSegment.isPositive());
+				if (nextCurrentSegment == null) { 
+					// end of route (-1 or length + 1)
+					directionAllowed = false;
+				} else {
+					if (nextCurrentSegment.isSegmentAttachedToStart()) {
+						// TODO count how many times in (should we optimize processOneRoadIntersection for this case)
```

Comment:
TODO

Reply:
200-300 NANOseconds

Reply:
Count how many %, counts and total seconds